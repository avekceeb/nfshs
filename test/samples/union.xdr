/*
 * Types of files:
 */
enum filekind {
   TEXT = 0,       /* ascii data */
   DATA = 1,       /* raw data   */
   EXEC = 2        /* executable */
};

/*
 * File information, per kind of file:
 */
union filetype switch (filekind kind) {
  case TEXT:
     void;                   /* no extra information */
  case DATA:
     string creator<10>;     /* data creator         */
  case EXEC:
     string interpretor<42>; /* program interpretor  */
  default:
     foo bar;
};
