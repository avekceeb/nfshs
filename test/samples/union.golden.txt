( [ Left
      (TypeDefEnum
         (Identifier "filekind")
         (( Identifier "TEXT" , Left (DecConstant 0) ) :|
            [ ( Identifier "DATA" , Left (DecConstant 1) )
            , ( Identifier "EXEC" , Left (DecConstant 2) )
            ]))
  , Left
      (TypeDefUnion
         (Identifier "filetype")
         UnionBody
           { unionDiscriminant = DiscriminantEnum (Identifier "kind")
           , unionArms =
               CaseSpec
                 { caseSpecValues = Right (IdentifierRef "TEXT") :| []
                 , caseSpecDeclaration = DeclarationVoid
                 } :|
                 [ CaseSpec
                     { caseSpecValues = Right (IdentifierRef "DATA") :| []
                     , caseSpecDeclaration =
                         DeclarationString
                           (Identifier "creator") (Just (Left (DecConstant 10)))
                     }
                 , CaseSpec
                     { caseSpecValues = Right (IdentifierRef "EXEC") :| []
                     , caseSpecDeclaration =
                         DeclarationString
                           (Identifier "interpretor") (Just (Left (DecConstant 42)))
                     }
                 ]
           , unionDefault =
               Just
                 (DeclarationSingle
                    (TypeIdentifier (IdentifierRef "foo")) (Identifier "bar"))
           })
  ]
, ParserState
    []
    [ ( SourcePos
          { sourceName = "test/samples/union.xdr"
          , sourceLine = Pos 13
          , sourceColumn = Pos 1
          }
      , ( Identifier "filetype"
        , TypeUnion
            UnionBody
              { unionDiscriminant = DiscriminantEnum (Identifier "kind")
              , unionArms =
                  CaseSpec
                    { caseSpecValues = Right (IdentifierRef "TEXT") :| []
                    , caseSpecDeclaration = DeclarationVoid
                    } :|
                    [ CaseSpec
                        { caseSpecValues = Right (IdentifierRef "DATA") :| []
                        , caseSpecDeclaration =
                            DeclarationString
                              (Identifier "creator") (Just (Left (DecConstant 10)))
                        }
                    , CaseSpec
                        { caseSpecValues = Right (IdentifierRef "EXEC") :| []
                        , caseSpecDeclaration =
                            DeclarationString
                              (Identifier "interpretor") (Just (Left (DecConstant 42)))
                        }
                    ]
              , unionDefault =
                  Just
                    (DeclarationSingle
                       (TypeIdentifier (IdentifierRef "foo")) (Identifier "bar"))
              }
        )
      )
    , ( SourcePos
          { sourceName = "test/samples/union.xdr"
          , sourceLine = Pos 21
          , sourceColumn = Pos 6
          }
      , ( Identifier "bar" , TypeIdentifier (IdentifierRef "foo") )
      )
    , ( SourcePos
          { sourceName = "test/samples/union.xdr"
          , sourceLine = Pos 4
          , sourceColumn = Pos 1
          }
      , ( Identifier "filekind"
        , TypeEnum
            (( Identifier "TEXT" , Left (DecConstant 0) ) :|
               [ ( Identifier "DATA" , Left (DecConstant 1) )
               , ( Identifier "EXEC" , Left (DecConstant 2) )
               ])
        )
      )
    ]
    (fromList [] :| [ fromList [] ])
)